index.html

```diff
@@ -702,27 +712,150 @@ <h3>Validating an ExposedThingInit</h3>
   <section> <h3>The <dfn>discover()</dfn> method</h3>
     <pre class="idl">
       partial namespace WOT {
-        ThingDiscovery discover(optional ThingFilter filter = null);
+        Promise&lt;ThingDiscoveryProcess&gt; discover(optional ThingFilter filter = null);
       };
     </pre>
     <div>
       Belongs to the <a>WoT Discovery</a> conformance class. Starts the discovery process that will provide {{ThingDescription}} objects for <a>Thing Description</a>s that match an optional |filter:ThingFilter| argument of type {{ThingFilter}}. The method MUST run the following steps:
       <ol>
         <li>
-          If invoking this method is not allowed for the current scripting context for security reasons, [= exception/throw =] a {{"SecurityError"}} and abort these steps.
+          Return a {{Promise}} |promise:Promise| and execute the next steps [=in parallel=].
+        </li>
+        <li>
+          If invoking this method is not allowed for the current scripting context for security reasons, reject |promise| with a {{SecurityError}} and abort these steps.
         </li>
         <li>
-          Construct a <a>ThingDiscovery</a> object |discovery:ThingDiscovery| with |filter|.
+          If discovery is not supported by the implementation, reject |promise|
+          with {{NotSupportedError}} and abort these steps.
         </li>
         <li>
-          Invoke the <a href="#the-start-method">discovery.start()</a> method.
+          Let |discovery:ThingDiscoveryProcess| be a new {{ThingDiscoveryProcess}} object.
         </li>
         <li>
-          Return |discovery|.
+          Set |discovery|'s {{ThingDiscoveryProcess/[[filter]]}} to |filter:ThingFilter|.
+        </li>
+        <li>
+          Set |discovery|'s {{ThingDiscoveryProcess/[[url]]}} to `undefined`.
+        </li>
+        <li>
+          If |filter| is not supported by the implementation, or if discovery
+          cannot be started by the underlying platform,
+          reject |promise| with {{NotSupportedError}} and abort these steps.
```

Comment:
I would split the two cases into two different errors. Throwing `NotSupportedError` for a condition in which the discovery cannot be started feels unspecific. Shouldn't be better to throw https://webidl.spec.whatwg.org/#operationerror ?

Reply:
Right. And there could be even 3 errors, 
- an error if the filter is not supported at all,  (`NotSupportedError`?)
- another if the filter contained an error, (`DataError`?)
- and yet another if discovery could not be started (`OperationError`?)
It's also a balance between being specific and giving fingerprinting differentiation. 
If we need one error, maybe `OperationError` would be fine.

Reply:
Maybe `NotSupportedError` and `OperationError` would be enough. I think the information that a filter is not supported, or filters in general are not supported, is not directly usable for fingerprinting, though correlations with HW capability, impl version etc could be statistically made. It's a policy decision that can even alter request by request or filter by filter.

The question is what to do if there is an error in a filter. Should the algoritm
- ignore the filter and provide everything it gets
- fail. 
Currently it fails and I think it should.